File: src/python/array/three-sum-closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        ret = None
        if nums is None or len(nums) <= 2:
            return ret
        nums.sort()
        for i, num in enumerate(nums):
            left = i+1
            right = len(nums)-1
            while left<right:
                val = num + nums[left] + nums[right]
                if ret is None or abs(ret-target) >= abs(val-target):
                    ret = val
                    if val >= target:
                        right -= 1
                    else:
                        left += 1
                elif val <= target:
                    left += 1
                elif val > target:
                    right -= 1
        return ret

File: src/medium/array/ThreeSumClosest.java
package src.medium.array;

import java.util.Arrays;

public class ThreeSumClosest {
    public int threeSumClosest(int[] nums, int target) {
      if (nums==null || nums.length<3)  return Integer.MAX_VALUE;
        
      Arrays.sort(nums);
      int ret = nums[0]+nums[1]+nums[num.length-1];
      
      for (int i=0; i<nums.length-2; i++){
          int j=i+1, k=nums.length-1;
          while(j<k){
              int temp = nums[i]+nums[j]+nums[k];
              if (temp<target){
                  ret = Math.abs(temp-target)<Math.abs(ret-target)?temp:ret;
                  j++;
              }
              else if (temp>target){
                  ret = Math.abs(temp-target)<Math.abs(ret-target)?temp:ret;
                  k--;
              }
              else
                  return target;
          }
      }
      
      return ret;
    }
}